performance/autoupdate-clients/main.go

```diff
@@ -0,0 +1,258 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"log"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"os"
+	"os/signal"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+)
+
+func main() {
+	N := getNFromCommandLine()
+	host, user, pw := getParams()
+	jar := getCookieJar(user, pw, host)
+
+	sumBytes := 0
+	oldSumKB := 0
+	changeIds := make(map[int]int)
+	changeIdsMutex := &sync.Mutex{}
+	gotUpdate := false
+
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	quit := make(chan int)
+	recv := make(chan Message)
+	for i := 0; i < N; i++ {
+		go client(quit, recv, i, jar, host)
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	go func(quit chan int) {
+		for {
+			select {
+			case <-quit:
+				return
+			case <-time.After(time.Second):
+				if !gotUpdate {
+					continue
+				}
+				gotUpdate = false
+
+				// count the amount of change ids per change id
+				amountChangeIds := make(map[int]int)
+				changeIdsMutex.Lock()
+				for _, changeId := range changeIds {
+					if _, ok := amountChangeIds[changeId]; ok {
+						amountChangeIds[changeId] += 1
+					} else {
+						amountChangeIds[changeId] = 1
+					}
+				}
+				changeIdsMutex.Unlock()
+
+				// get size of update
+				sumKB := int(sumBytes / 1024)
+				diffKB := sumKB - oldSumKB
+				oldSumKB = sumKB
+				log.Printf("New update: %d KB (diff %d KB)\n", sumKB, diffKB)
+
+				sortedChangeIds := make([]int, len(amountChangeIds))
+				i := 0
+				for changeId := range amountChangeIds {
+					sortedChangeIds[i] = changeId
+					i++
+				}
+				sort.Ints(sortedChangeIds)
+				for _, changeId := range sortedChangeIds {
+					log.Printf("%d: %d\n", changeId, amountChangeIds[changeId])
+				}
+
+				// change id delta: difference between min and max change id
+				delta := sortedChangeIds[len(sortedChangeIds)-1] - sortedChangeIds[0]
+				log.Printf("delta=%d\n", delta)
+
+				log.Printf("\n")
+			}
+		}
+	}(quit)
+
+	for {
+		select {
+		case <-interrupt:
+			for i := 0; i < (N + 1); i++ {
+				quit <- 0
+			}
+			return
```

Comment:
This will stop the program immediately. If you do not wait for the goroutines to quit, then you don't need the quit channel at all.